marshal/world.go

```diff
@@ -297,6 +296,7 @@ func (m *Marshaler) ClaimPartition(topicName string, partID int) bool {
 // still owning this partition. Returns an error if anything has gone wrong (at which
 // point we can no longer assert we have the lock).
 func (m *Marshaler) Heartbeat(topicName string, partID int, lastOffset int64) error {
+	// cannot be called within topic.lock.RLock() since it calls topic.lock.Lock()
```

Comment:
yeah..it was a bit painful....there are quite a few locks and the problem with Go is that locks are not re-entrant, which is a bit annoying....I had to write code that dump stacks to figure out what's going on.